client/client.go

```diff
@@ -0,0 +1,368 @@
+	}
+	sink = make(chan *bindings.AccountantImplementationPromiseSettled)
+	addr, err := bc.getProviderChannelAddressBytes(accountantID, providerID)
+	if err != nil {
+		return sink, cancel, errors.Wrap(err, "could not get provider channel address")
+	}
+	sub, err := caller.WatchPromiseSettled(&bind.WatchOpts{}, sink, [][32]byte{addr})
+	if err != nil {
+		return sink, cancel, errors.Wrap(err, "could not subscribe to promise settlement events")
+	}
+
+	go func() {
+		subErr := <-sub.Err()
+		if subErr != nil {
+			log.Error().Err(err).Msg("subscription error")
+		}
+		close(sink)
+	}()
+
+	return sink, sub.Unsubscribe, nil
+}
+
+// IsRegistered checks wether the given identity is registered or not
+func (bc *Blockchain) IsRegistered(registryAddress, addressToCheck common.Address) (bool, error) {
+	caller, err := bindings.NewRegistryCaller(registryAddress, bc.client)
+	if err != nil {
+		return false, errors.Wrap(err, "could not create registry caller")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), bc.bcTimeout)
+	defer cancel()
+
+	res, err := caller.IsRegistered(&bind.CallOpts{
+		Context: ctx,
+	}, addressToCheck)
+	return res, errors.Wrap(err, "could not check registration status")
+}
+
+// GetMystBalance returns myst balance
+func (bc *Blockchain) GetMystBalance(mystAddress, identity common.Address) (*big.Int, error) {
+	c, err := bindings.NewMystTokenCaller(mystAddress, bc.client)
+	if err != nil {
+		return nil, err
+	}
+	parent := context.Background()
+	ctx, cancel := context.WithTimeout(parent, bc.bcTimeout)
+	defer cancel()
+	return c.BalanceOf(&bind.CallOpts{
+		Context: ctx,
+	}, identity)
+}
+
+// GetRegistrationFee returns fee required by registry
+func (bc *Blockchain) GetRegistrationFee(registryAddress common.Address) (*big.Int, error) {
+	// TODO to reduce amount of blockchain calls, it could get registration fee from cache (updated once in a day)
+	c, err := bindings.NewRegistryCaller(registryAddress, bc.client)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get registration fee")
+	}
+	parent := context.Background()
+	ctx, cancel := context.WithTimeout(parent, bc.bcTimeout)
+	defer cancel()
+	res, err := c.RegistrationFee(&bind.CallOpts{
+		Context: ctx,
+	})
+	return res, errors.Wrap(err, "could not get registration fee")
+}
+
+// RegistrationRequest contains all the parameters for the registration request
+type RegistrationRequest struct {
+	WriteRequest
+	AccountantID    common.Address
+	Loan            *big.Int
+	Fee             *big.Int
+	Beneficiary     common.Address
+	Signature       []byte
+	RegistryAddress common.Address
+}
+
+// WriteRequest contains the required params for a write request
+type WriteRequest struct {
+	Identity common.Address
+	Signer   bind.SignerFn
+	GasLimit uint64
+	GasPrice *big.Int
+}
+
+// RegisterIdentity registers the given identity on blockchain
+func (bc *Blockchain) RegisterIdentity(rr RegistrationRequest) (*types.Transaction, error) {
```

Comment:
Do you have reasons when to pass value or pointer and why? As an example here I see that RegistrationRequest is passed by value but it returns pointer to types.Transaction. Is some other places it returns value etc.

In my opinion struct like RegistrationRequest could be passed by pointer as it contains a lot of fields and some of them are already pointers so it will probably will not escape heap any way, also struct is not needed to be copied. For such RPC like clients I suggest to be consistent use pointers for both requests and responses.

Reply:
I pass everything as value not by reference as I have no intention of modifying the existing struct. This is a good indicator to the consumer of the API.